// @module-federation/enhanced@0.21.3/dist/src/lib/container/runtime/EmbedFederationRuntimeModule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
    MIT License http://www.opensource.org/licenses/mit-license.php
    Author Zackary Jackson @ScriptedAlchemy
*/
const normalize_webpack_path_1 = require("@module-federation/sdk/normalize-webpack-path");
const { RuntimeModule, Template, RuntimeGlobals } = require((0, normalize_webpack_path_1.normalizeWebpackPath)('webpack'));
class EmbedFederationRuntimeModule extends RuntimeModule {
    constructor(containerEntrySet) {
        super('embed federation', RuntimeModule.STAGE_ATTACH);
        this.containerEntrySet = containerEntrySet;
        this._cachedGeneratedCode = undefined;
    }
    identifier() {
        return 'webpack/runtime/embed/federation';
    }
    generate() {
        if (this._cachedGeneratedCode !== undefined) {
            return this._cachedGeneratedCode;
        }
        const { compilation, chunk, chunkGraph } = this;
        if (!chunk || !chunkGraph || !compilation) {
            return null;
        }
        let found;
        if (chunk.name) {
            for (const dep of this.containerEntrySet) {
                const mod = compilation.moduleGraph.getModule(dep);
                if (mod && compilation.chunkGraph.isModuleInChunk(mod, chunk)) {
                    found = mod;
                    break;
                }
            }
        }
        if (!found) {
            return null;
        }
        const initRuntimeModuleGetter = compilation.runtimeTemplate.moduleRaw({
            module: found,
            chunkGraph,
            request: found.request,
            weak: false,
            runtimeRequirements: new Set(),
        });
        const result = Template.asString([
            `var prevStartup = ${RuntimeGlobals.startup};`,
            `var hasRun = false;`,
            `${RuntimeGlobals.startup} = ${compilation.runtimeTemplate.basicFunction('', [
                `if (!hasRun) {`,
                `  hasRun = true;`,
                `  ${initRuntimeModuleGetter};`,
                `}`,
                `if (typeof prevStartup === 'function') {`,
                `  return prevStartup();`,
                `} else {`,
                `  console.warn('[Module Federation] prevStartup is not a function, skipping startup execution');`,
                `}`,
            ])};`,
        ]);
        this._cachedGeneratedCode = result;
        return result;
    }
}
exports.default = EmbedFederationRuntimeModule;
