// @modern-js/utils@2.68.2/dist/esm/compiled.js
import { Import } from "./import";
import { default as default2 } from "../compiled/fs-extra";
import { default as default3 } from "../compiled/ora";
import { default as default4 } from "../compiled/glob";
import { default as default5 } from "../compiled/js-yaml";
import { default as default6 } from "../compiled/chalk";
import { default as default7 } from "../compiled/debug";
import { default as default8 } from "../compiled/slash";
import { default as default9 } from "../compiled/execa";
import { default as default10 } from "../compiled/json5";
import { default as default11 } from "../compiled/upath";
import { default as default12 } from "../compiled/pkg-up";
import { nanoid } from "../compiled/nanoid";
import { default as default13 } from "../compiled/semver";
import { default as default14 } from "../compiled/dotenv";
import { default as default15 } from "../compiled/lodash";
import { default as default16 } from "../compiled/globby";
import { default as default17 } from "../compiled/address";
import { default as default18 } from "../compiled/signale";
import { default as default19 } from "../compiled/url-join";
import { default as default20 } from "../compiled/minimist";
import { default as default21 } from "../compiled/fast-glob";
import { default as default22 } from "../compiled/filesize";
import { default as default23 } from "../compiled/gzip-size";
import { default as default24 } from "../compiled/strip-ansi";
import { default as default25 } from "../compiled/dotenv-expand";
import { default as default26 } from "../compiled/browserslist";
import { program, Command } from "../compiled/commander";
import { Signale } from "../compiled/signale";
var mime = Import.lazy("../compiled/mime-types", require);
var chokidar = Import.lazy("../compiled/chokidar", require);
var inquirer = Import.lazy("../compiled/inquirer", require);
export {
  Command,
  Signale,
  default17 as address,
  default26 as browserslist,
  default6 as chalk,
  chokidar,
  default7 as debug,
  default14 as dotenv,
  default25 as dotenvExpand,
  default9 as execa,
  default21 as fastGlob,
  default22 as filesize,
  default2 as fs,
  default4 as glob,
  default16 as globby,
  default23 as gzipSize,
  inquirer,
  default10 as json5,
  default15 as lodash,
  mime,
  default20 as minimist,
  nanoid,
  default3 as ora,
  default12 as pkgUp,
  program,
  default13 as semver,
  default18 as signale,
  default8 as slash,
  default24 as stripAnsi,
  default11 as upath,
  default19 as urlJoin,
  default5 as yaml
};
